add v2

tolsen opened the pull request:
Includes Bard's changes and RollingFileAppender

Commit:
copy v1 to create v2

Commit:
Apply Bard's changes.

This adds the ability to ignore certain filenames when logging the
caller (works well when building a logging system on top of slogger)

Also adds ROUTINE and DOOM log levels.

Commit:
Disable cache in v2 as it's not fully implemented and the part that is
implemented has locking overhead

Fix v2 logger_test to use v2 path

Commit:
Add rolling_file_appender.go and test

Commit:
sort logHeader into proper place

Commit:
refactor creation of Log's in rollling_file_appender.go

dgottlieb commented on v2/logger.go:
```diff
@@ -0,0 +1,207 @@
+
+func nonSloggerCaller() (pc uintptr, file string, line int, ok bool) {
+	for skip := 0; skip < 100; skip++ {
+		a,b,c,ok := runtime.Caller(skip)
```
`a, b, c, ok` should be `pc, file, line, ok`

tolsen replied:
cc'ing @bardbloom as this is originally his change.  Agreed, in particular as pc, file, and line are in the function signature.  I'll make the change

tolsen replied:
done

dgottlieb commented on v2/logger.go:
```diff
@@ -0,0 +1,207 @@
+	for skip := 0; skip < 100; skip++ {
+		a,b,c,ok := runtime.Caller(skip)
+		shouldBeIgnored := containsAnyIgnoredFilename(b)
+		if !ok || !shouldBeIgnored  {
```
I would instead check the `ok` value from `runtime.Caller` before bothering to call `containsAnyIgnoredFilename`

tolsen replied:
cc'ing @bardbloom again.  Yes, in fact if we remove the shouldBeIgnored variable and move the containsAnyIgnoreFilename call into the if, the short-circuiting || operator will have the same effect.

tolsen replied:
done

bardbloom-zz replied:
Heh.  At one point there was probably a println in between those lines, or
I thought there would soon be one.

On Mon, Aug 19, 2013 at 10:16 AM, Tim Olsen notifications@github.comwrote:

> In v2/logger.go:
> 
> > +}
> > +
> > +func containsAnyIgnoredFilename(s string) bool {
> > -   for _, ign := range ignoredFileNames {
> > -       if strings.Contains(s, ign)  {
> > -           return true
> > -       }
> > -   }
> > -   return false
> >   +}
> >   +
> >   +func nonSloggerCaller() (pc uintptr, file string, line int, ok bool) {
> > -   for skip := 0; skip < 100; skip++ {
> > -       a,b,c,ok := runtime.Caller(skip)
> > -       shouldBeIgnored := containsAnyIgnoredFilename(b)
> > -       if !ok || !shouldBeIgnored  {
> 
> cc'ing @bardbloom https://github.com/bardbloom again. Yes, in fact if
> we remove the shouldBeIgnored variable and move the
> containsAnyIgnoreFilename call into the if, the short-circuiting ||
> operator will have the same effect.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/10gen-labs/slogger/pull/1/files#r5842933
> .

dgottlieb commented on v2/logger.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+func containsAnyIgnoredFilename(s string) bool {
+	for _, ign := range ignoredFileNames {
```
The code does not do what the documentation says it does. The documentation/API method implies just ignoring filenames (i.e. not directories), but the code ignores on substrings of the full, absolute file path.

tolsen replied:
cc'ing @bardbloom.  How about ignoredPathSubstrs, IgnoreThisPathSubstrToo (not sure if the Too is needed), and containsIgnoredPathSubstr ?  Other suggestions also welcome

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+
+func (self RollingFileAppender) Append(log *Log) error {
+	select {
+	case self.appendCh <- log:
```
Suggestion: Instead of immediately logging a `fullWarningLog` on every channel full, one can do something like:

```
select {
case self.appendCh <- log:
  //nothing
case <-time.After(5 * time.Second):
  self.appendCh <- fullWarningLog()
  self.appendCh <- log
}
```

I expect a small wait (5 seconds was only an example) would clean up a few instances of multi-logging a full warning log.

tolsen replied:
I think there should be a warning the moment the channel is full so the user is aware any time the logging system is not able to be asynchronous.  That said, I can see the desire not to log it multiple times.  I can try to implement a solution that will log it the first time, but not log it again until the channel is has gone below 90% full and fills up again.  What do you think?

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+func (self RollingFileAppender) listenForAppends() {
+	needsSync := false
+	for {
+		if needsSync {
```
I'm not positive, but this if statement might be collapsable with the introduction of a single select that takes advantage of `<-time.After(duration)`.

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+
+func (self RollingFileAppender) Sync() {
+	self.syncCh <- true
+	for !(<- self.syncCh) {
```
My interpretation of what is going on here might be wrong, but this feels like a spin lock where this while loop is oftentimes reading the value the `Sync()` method added to itself. Edit: I am reading this incorrectly, this channel is a synchronous channel.

The semantics of the method itself are a little weird. It's not abundantly clear that this blocks until the logging channel is empty as opposed to some useful notion of "Block until all logs issued before this invocation are persisted to disk, but not necessarily any of the logs instantiated after".

A third point, using the `Sync()` signature as a name collides with a method on the embedded os.File. If changing this signature of this method in the future, the compiler won't catch any problems and users will unwittingly be telling the file to flush to disk.

I suggest changing the name to say `WaitUntilEmpty` to kill 2 and 3 with 1 stone.

tolsen replied:
per this comment and our in-person discussion, I've renamed Sync to waitUntilEmpty() , thereby also making it private.  This is so that it's only used by Close() which is intended to be used when logging is dying down.

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+const ROLLING_FILE_APPENDER_CHANNEL_SIZE = 4096
+
+type RollingFileAppender struct {
+	*os.File
```
Given the rolling appender writes to different files over time, this feels wrong to be an embedded type.

tolsen replied:
make sense. will change.

tolsen replied:
s/make/makes/

tolsen replied:
Done.  I've also changed NewRollingFileAppender() to take a path string (does not have to be absolute) instead of a File.  NewRollingFileAppender() will figure out the absolute path of the path string if it's not already an absolute path so as to prevent problems if the working directory changes later.

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+	
+	// close current log
+	err := self.File.Close()
+	if err != nil {
```
There is a variation of the if statement in go that has a "precondition":

```
if err := self.File.Close(); err != nil {
  self.errHandler(...)
}
```

tolsen replied:
done

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+	return
+}
+
+// returns true on success, false otherwise
```
Not sure if there's a benefit to returning a boolean instead of nil/error.

tolsen replied:
The implication is that any errors have already been "handled" in that an error has been sent to the error handler.  What do you think?  Do you prefer to return err in this case?

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+	}
+
+	// rename old log
+	if !self.renameLogFile(filename, newRotatedFilename(filename)) {
```
Does this process of renaming files preserve the directory the logs are being written to assuming the input `os.File` to the constructor is not the same as the current working directory?

dgottlieb replied:
Also:

If a rename fails, the the appender tries to re-open the old file. If that fails, it sets File to nil and returns false. On the return of false, this function also returns. None of the error messages in this execution path imply anything about future logging failing (as the below error message does).

tolsen replied:
1. should now be addressed by the change to record the absolute path at construction
2. good catch.  I'll refactor the error message to be on both paths

tolsen replied:
Actually, I've just modified the "Error while reopening %s after failing to rename." error message to be "Error while reopening %s after failing to rename. Further logging will fail".

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+	if err != nil {
+		self.File = nil
+		self.errHandler(fmt.Errorf(
+			"Failed to create %s . Further logging wil fail. : %s",
```
will\* fail

tolsen replied:
fixed

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+	curFileSize uint64
+	appendCh chan *Log
+	syncCh chan bool
+	errHandler func(error)
```
(Not sure where to place this comment)

Given there's an error handler callback which makes sense due to the multi-threaded nature of how logging errors will be discovered, I would suggest creating a safe way to determine the error (as opposed to parsing a string). For example, the error that fails to create a new log file is pretty important to deal with (I'm assuming the others are less-so) and the application `errHandler` would have to parse a string to determine what the error is. If the string changes, existing `errHandlers` will break.

dgottlieb replied:
Adding onto the last thought about the error handling in general: the user application really loses any sensible way of understanding what is going wrong. The error handler should perhaps be a `func(rollingFileError, error)` to give full context of what part of the logger failed and the underlying cause to that failure.

tolsen replied:
I've now added several structs that implement error.  I'm leaving the errHandler signature as is, but the user can call IsCloseError(err) for example to figure out if it is a CloseError or they can do a type switch.

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+	n, err := self.File.WriteString(msg)
+
+	if err != nil {
+		self.errHandler(fmt.Errorf("Could not log to %s : %s", self.File.Name(), err.Error()))
```
If a log failed to write to a file, I'm not sure if this code should continue to attempt to track the size and potentially rotate.

tolsen replied:
fixed to return early

dgottlieb commented on v2/rolling_file_appender.go:
```diff
@@ -0,0 +1,233 @@
+
+func (self RollingFileAppender) reallyAppend(log *Log, trackSize bool) {
+	if self.File == nil {
+		self.errHandler(errors.New("I have no logfile to write to!"))
```
This probably needs to return before trying to dereference a nil pointer below?

tolsen replied:
good catch.  fixed to return early

dgottlieb commented:
Two notes:

First, was there any intention of offering a way to delete old log files either via "keep around the last X files" or "delete files older than X days"?

Second, the error handling, while probably unlikely to ever really come up, feels very incomplete. If the rolling file appender did synchronous writes (say through a channel of channels) the errors could be directly bubbled up through each call to a log method. If we want to keep application logging and writing to the file asynchronous, we should brainstorm some simple, common policies that the `errorHandler` might want to implement and ensure the API is sufficient for doing so.

Commit:
use better named variables for return values from runtime.Caller()

Commit:
In nonSloggerCaller(), check ok from runtime.Caller() before checking
for ignored filenames

Commit:
rename (RollingFileAppender) Sync() to waitUntilEmpty().  This also
makes it private

Commit:
Change NewRollingFileAppender to take a path instead of a File.  The
resulting File is no longer embedded but is stored as a field name
file.

Commit:
have (RollingFileAppender) Close() return error instead of nothing

Commit:
tighten up Close() call and error check in (RollingFileAppender) rotate()

Commit:
indicate in error message that further logging will fail if we fail to
reopen the log file after failing to rotate

Commit:
Fix error message typo

Commit:
Have reallyAppend() return early if file is nil

Commit:
fix reallyAppend() to return early if there is an error while logging

Commit:
have RollingFileAppender log header upon construction regardless if it's a new file

Commit:
Change syncCh to be a chan (chan bool) so that a separate channel is
used for replies so that multiple simultaneous calls to Close() do not
cause problems.

Commit:
move rolling_file_appender into its own subpackage

Make messageFmt and args in Log struct public

Commit:
shorten NewRollingFileAppender() to New() and
ROLLING_FILE_APPENDER_CHANNEL_SIZE to APPEND_CHANNEL_SIZE

Commit:
add CloseError, NoFileError, OpenError, RenameError, and WriteError to
rolling_file_appender and pass one of them to errHandler upon error

tolsen commented:
I've addressed the error handling.  I'll work on limiting the number of log files now.

Commit:
Simplify Is*Error() methods to not check for pointer types

Commit:
Add feature to limit number of rotated log files

tolsen commented:
Old log files are now deleted.  @dgottlieb , that should be everything.

Commit:
add TestLog and TestConcurrentLog to rolling_file_appender_test

Commit:
increment second field of log file name by 1 when rotating if such a
log file already exists

Commit:
Have rolling_file_appender pull in slogger via non local import.  We
will have to change this after the pull request in the main slogger
repo.

Commit:
Disable auto log rotation if a negative maxFileSize is passed on
construction

Commit:
add rotateIfExists option to rolling_file_appender.New()

Commit:
change headerGenerator callback to return []string

Conflicts:
	v2/rolling_file_appender/rolling_file_appender.go
	v2/rolling_file_appender/rolling_file_appender_test.go

Commit:
Add newLevel(string)Level function.

Commit:
make newLevel public

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+	"regexp"
+	"runtime"
+	"time"
+	"github.com/tolsen/slogger/v2"
```
Should this be `github.com/10gen-labs/slogger/v2`?

tolsen replied:
Unfortunately, not until this code is merged into 10gen-labs.  Go allows you to have a relative path such as ".." but then that fails when you import the library into another project.   This makes it really inconvenient to work on separate branches or repositories.  It seems that all forks and merges would need to be followed up by a commit to change this import.  Any ideas?

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+}
+
+// Set maxFileSize to < 0 for unlimited file size (no rotation)
+func New(filename string, maxFileSize int64, maxRotatedLogs int, rotateIfExists bool, errHandler func(error), headerGenerator func() []string) (*RollingFileAppender, error) {
```
Probably nice to document this method as it's the most important for the user. Details worth noting in the documentation:
- How is `filename` handled? What log file gets written to, what do rotated files look like.
- Explain `rotateIfExists`. If `rotateIfExists` is false, does it overwrite or append to the current file?
- `errHandler` can be nil. What kind of errors can pop up as input to the callback?
- What does the `headerGenerator` do?

tolsen replied:
done

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+// Set maxFileSize to < 0 for unlimited file size (no rotation)
+func New(filename string, maxFileSize int64, maxRotatedLogs int, rotateIfExists bool, errHandler func(error), headerGenerator func() []string) (*RollingFileAppender, error) {
+	if errHandler == nil {
+		errHandler = func(err error) { }
```
I find it inconsistent that an empty error handler is instantiated in the constructor, while the `headerGenerator` is nil-checked only when it's about to be called. I think it would be more straightforward to instantiate an empty header generator in the constructor as well. Looking at this function (in lieu of documentation) I feel I can pass `nil` for the `errHandler` but I must give a valid `headerGenerator`.

tolsen replied:
I have no problem making this consistent.  I originally did it this way because the number of call sites for the errHandler is high enough that I would like to avoid a nil check at each one.  The headerGenerator, however, is only called in one place and I assumed that a nil check is lower overhead than an function call to a stub.  But I have no problem with consistency trumping that.

tolsen replied:
done

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+		headerGenerator: headerGenerator,
+	}
+
+	if rotateIfExists {
```
I had some trouble parsing what this big if-else block was trying to accomplish. Might be worth writing this as three explicit cases being (not sure of this proposal myself, but I feel there is room for improvement):
- The file does not exist, the `rotateIfExists` flag is irrelevant.
- The file exists and we want to rotate.
- The file exists and we want to append.

It may be preferable to even consolidate a file not existing and appending to a file as the same general case.

tolsen replied:
good eye.  I was able to reduce the code size to less than 1/3 the original and remove the second Stat() call!  committed & pushed

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+	)
+}
+
+func internalWarningLog(messageFmt string, args []interface{}) *slogger.Log {
```
To take advantage of how `Printf` or `Logf` works with "vararg" syntax:

```
func (self *Logger) Logf(level Level, messageFmt string, args ...interface{}) (*Log, []error) {
  return self.logf(level, messageFmt, args...)
}
```

tolsen replied:
done

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+
+func (self *RollingFileAppender) listenForAppends() {
+	needsSync := false
+	for {
```
Given all the magic happens here, I can see this lots of eyes finding their way to this `while` loop; it would be nice to document the expected behavior.

Also, there was a deadlocking happening with one of the goroutines in this loop, right? Did you find out why that was happening? Would be personally useful to help avoid running into unexpected deadlocks.

tolsen replied:
documented.

The deadlock was occurring because of a silent panic in the listenForAppends() goroutine.  If I remember correctly, "go test" appears to be silent about goroutines dying.

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+		}
+	}
+	return
+}
```
Is this fix platform specific? Might want to test on some 32-bit system. I would like to once again emphasize the less fragile approach of using the first value in the list of being the current "oldestTime" in the below loop iteration.

tolsen replied:
Ok, but rather than using the first value as the initial oldestTime, I've added a foundValidTime boolean.  This way I don't have to add a second for loop to the find the first valid time

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+	// find oldest Time
+	var oldestTime time.Time = maxTime
+	for _, timeStr := range timeStrs {
+		rotatedTime, err := time.Parse("2006-01-02T15-04-05", timeStr)
```
I believe the case of `err != nil` should be handled, particularly if the layout does not follow in lock step with the time format being written out. In that case old files will never be deleted?

tolsen replied:
The concern here is that a user may copy a logfile, at which point we may not want to delete it.  For example, if I copy "log.2006-01-02T15-04-05" to "log.2006-01-02T15-04-05-copy" then I might not want the program to remove it out from under me

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+	}
+
+	// recursively call ourself if there's more to do
+	if len(timeStrs) > self.MaxRotatedLogs {
```
TCO: Tail-Recursive optimization? TRO? May want to spell out the acronym.

Edit: Ahh it's tail call optimization.

tolsen replied:
spelled out

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+	// Now we are in a weird case where we were over the limit by more
+	// than one.  Rather than complicate the above code to find the N
+	// oldest times we will just recursively call ourselves, but only
+	// if we have made any progess to avoid going into an infinite
```
I think making this non-recursive complicates the code only because there is no separation between policy (when to delete the oldest log file) and mechanism (a function that will find and remove the oldest log file without any checks about when to do it). Under this factoring, code can then look like:

```
func removeOldestLog(...) error {
  // Return an error if there is no log identified to be deleted or the delete fails?
  // All the same fun stuff as above with regards to getting
  // the oldest file and deleting
}

func removeMaxRotatedLogs() {
  timeStrs, err := self.rotatedTimeStrs()
  if err != nil {
    self.errHandler(MinorRotationError{err})
    return
  }

  filesToRemove := len(timeStrs) - self.MaxRotatedLogs
  for removed := 0; removed < filesToRemove; removed++ {
    if err = self.removeOldestLog(); err != nil {
      self.errHandler(MinorRotationError{err})
      return
    }
  }
}
```

tolsen replied:
I should have explained it better in the comment, but what I really meant was that I did not want to complicate the code with an algorithm that is better than O(n^2).  The code you give is also O(n^2).  Personally, I do not mind the tail call, particularly if I think it will be TCO'd.

tolsen replied:
Nevermind, I've decided to rewrite this anyway in light of the new serial number.  I will sort the files.  This means the code is now Theta(n log n) instead of Big-Omega(n) (best case) and O(n^2) (worst case).  The number of log files should be small so I am not too concerned with the best case being worse

tolsen replied:
hmm.. nevermind, I'm assuming performance characteristics of go's built-in sort.  oh well, I don't think it should matter much.

tolsen replied:
Ok, I've reimplemented removeMaxRotatedLogs to sort the logfiles and then remove the oldest ones.

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+func (self *RollingFileAppender) renameLogFile(oldFilename string, inc int) (ok bool) {
+	newFilename := newRotatedFilename(self.absPath, inc)
+	_, err := os.Stat(newFilename) // check if newFilename already exists
+	if err == nil {
```
I'm skeptical of this logic. Suppose one wants to keep 4 files, but in a single second, the appender wants to "roll over" 5 times. If the start time is at 10 seconds, we'll get files:

```
10 seconds
10 seconds renamed to 11 seconds
10 seconds renamed to 12 seconds
10 seconds renamed to 13 seconds
(delete 10 seconds file because it's being rolled over, or maybe that happens after the following file creation, but the point still holds)
10 seconds (despite being newer than 13 seconds)
```

I would consider using nano seconds to differentiate as needed, though I realize that breaks the recursive structure of the function. Even sleeping might be desirable; so long as there's room in the `appendCh` there should be no blocking?

tolsen replied:
I'm against sleeping because if the user is logging so fast that logs are rotating more than once per second then there is an increased risk that appendCh will fill up.  I don't mind breaking the recursive structure but I am concerned about changing the filename format by appending nanoseconds in the case of conflicts.  Or are you suggesting we always add nanoseconds?  Hmm... That would make it different than mongodb's rotated logfile format but maybe that's ok.

tolsen replied:
Reimplemented to append a serial number when there's a conflict

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+	err = os.Rename(oldFilename, newFilename)
+
+	
+	if err != nil {
```
Any reason why the old file is being re-opened as read/write whereas the constructor opens with write only and append?

dgottlieb replied:
As with the other file permissions, I'm curious about the write privileges for everyone.

tolsen replied:
Good catch.  This should be reopened w/ at least append and probably write-only as well.

As for write privileges for everyone, I just tested a small program and the permissions are masked by the umask.  This is the standard way to set permissions when creating a file if you want to leave it to the user's umask (although in this case the file should already exist).

tolsen replied:
ok, I've changed the reopening of the old log file to be O_WRONLY and O_APPEND

dgottlieb commented on v2/rolling_file_appender/rolling_file_appender.go:
```diff
@@ -0,0 +1,484 @@
+	}
+
+	return
+}
```
Don't think it helps to declare the `ok bool` variable if it's not being used.

tolsen replied:
It helps as a form of documentation.  Note that I removed the comment that used to say it returned whether it succeeded.

Commit:
change (RollingFileAppender *) Append() to not be on a pointer
receiver

Commit:
License slogger under the Apache License v2 and add copyright notices.

Commit:
Better document RollingFileAppender's New()

Commit:
Default headerGenerator to a default function that returns an empty
slice.  This removes the need for the nil check in logHeader()

Commit:
simplify file open logic in RotatingFileAppender's New()

Commit:
change internalWarningLog() to take varargs in the spirit of Printf,
etc.

Commit:
document RotatingFileAppender's listenForAppends()

Commit:
change removeMaxRotatedLogs to not rely on time.Before() working
correctly with large time values

Commit:
expand TCO acronym in comment

Commit:
Fix bug which would sometimes remove logfiles which were not
necessarily the oldest.

Simplify removeMaxRotatedLogs.  Introduce RotationTime and
RotationTimeSlice

Commit:
Have rolling_file_appender_test create its test directory with 777
permissions so we respect the umask

Commit:
Change reopen of old log file to be O_WRONLY and O_APPEND in case of
failed rename

tolsen commented:
Ok, that should be everything.

Commit:
Add Flush() to Logger and Appender

Commit:
rolling_file_appender: add package comment and add comment to New()
suggesting that a defer call to Close() (or Flush()) is made after
New()

Commit:
add queue sub-package

Commit:
add queued_set sub-package

Commit:
Add support to Logger for suppressing repeated logs

Commit:
Have *Logger.Flush() not return nils in the []error slice it returns.
If there are no errors, then it returns an empty slice.

Commit:
Change RollingFileAppender to be synchronous, in preparation for
factoring out the asynchronous part into AsyncAppender, and having a
new AsyncRollingFileAppender that composes the two.

Commit:
pull up rolling_file_appender.simpleLog() into slogger.SimpleLog()

Commit:
Add AsyncAppender

Commit:
Fix header comment in async_appender.go to refer to async_appender

Commit:
move assertNoErrors to new test_util package

Commit:
Fix async_appender.TestConcurrentLog to wait for logging goroutines to
finish.

Add some AssertNoErrors() checks to async_appender_test

Commit:
Add Context support

Commit:
Remove ability to pass context into Logf, Errorf, and Stackf
functions.  Instead, add LogfWithContext, ErrorfWithContext, and
StackfWithContext functions.

Commit:
set cache to nil when disabling log suppression so that the cache can
be garbage collected

Commit:
change order of setting cache to nil and setting suppressionEnabled to
false in DisableLogSuppression() in order to avoid period of time
where suppressionEnabled is true but cache is nil

Commit:
Nil out source location for header log lines

Commit:
Append should be on *AsyncAppender rather than AsyncAppender.

Commit:
Change args parameter to slogger.SimpleLog() to be variadic.  Add
slogger.SimpleLogStrippingDirs()

Commit:
Change order of arguments to SimpleLogStrippingDirs() so that
messageFmt is next to args

Commit:
Add Context.Keys()

Commit:
Add Level.String() method

Commit:
Have RollingFileAppender use an empty string for the filename and a
line number of 0 in the header.

Commit:
Alternative directory/package structure

Commit:
Adding support for TRACE log Level

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Fix RollingFileAppender to delete the oldest files (not the newest!!)

Commit:
Add function name to log message

Commit:
Strip off namespace from function name in log

Commit:
Update comment for queue regarding behavior when full.  Run go fmt on
queue.go

Commit:
Run go fmt on all go files

Commit:
Rename Log.cache to Log.suppressionCache

Commit:
Revert "Update comment for queue regarding behavior when full.  Run go fmt on"

This reverts commit eea0117439c5cf36fcbf968836d3b89c1a89376f.

Commit:
Run go fmt on queue.go

Commit:
Add RetainingLevelFilterAppender

Commit:
Add support for clearing retained logs

Commit:
Have RetainingLevelFilterAppender use 1 lock instead of 2 to prevent
possibility of deadlock.

Commit:
Introducing concept of turbo filters a la logback

Commit:
Adding clarifying doc comment on revised Logf behavior

Commit:
Merge pull request #2 from john-morales/feature_turbofilters

Introducing concept of turbo filters a la logback

Commit:
Initial commit

Commit:
Reintegration from 10gen-labs

Commit:
Reintegration from master

Commit:
Removing support for TRACE log level

This reverts commit a651a0675b6aa828eca4cc318b91e8f4622d9145.

Commit:
Add millisecond part of timestamp to log format

Commit:
During log rotation, close file before renaming it

dgolub committed:
Add support for custom StringWriter interface to handle writing lines to log files.

Commit:
Merge pull request #1 from dgolub/CM-1751

Add support for custom StringWriter interface to handle writing lines to log files.

leothekim committed:
Make tests work on Windows

Windows doesn't like it when multiple processes open the same file. This adds
defer calls to close the RollingFileAppender. Also, createLogDir calls
os.MkdirAll because os.Mkdir fails on Windows if the directory being created
already exists.

Commit:
Merge pull request #2 from leothekim/master

Make tests work on Windows

louisaberger committed:
Add ErrorCode first pass

Commit:
Merge pull request #3 from louisaberger/CLOUDP-12022

merging

louisaberger committed:
Wrap logged error in ErrorWithCode

Commit:
Merge pull request #4 from louisaberger/CLOUDP-12022c

Wrap logged error in ErrorWithCode

Commit:
Remove string interpolation support for contexts.  It was
unintentionally interpolating messages that contained prior formatted
messages that contained structs. (Example: &{1})

Commit:
Merge remote-tracking branch 'mongodb/master'

Commit:
CLOUDP-5982: Move slogger v2 code one-level down into a slogger
subdirectory

Commit:
CLOUDP-5982: Update import paths and tests for new directory layout

Commit:
CLOUDP-5982: Remove log suppression

Commit:
CLOUDP-5982: Remove ROUTINE logging level.  Rename DOOM logging level
to FATAL.

Commit:
CLOUDP-5982: Incorporate changes from v1 up to commit
9c88ac45b54cb0ca3359bb3c871cde90c50dd7d1

Commit:
CLOUDP-5982: Add missed files from update from v1

Commit:
Remove unused code and stale comment

Commit:
Remove log cache

Commit:
Pre-code review changes

Commit:
Add copyright and license to turbo filters

Commit:
Add copyright and license to test_util

Commit:
Update copyrights

Commit:
More copyright updates

Commit:
CLOUDP-5982: Make RollingFileAppender thread-safe

Commit:
CLOUDP-5909: Use OS independent os.PathSeparator in stripDirectories()
